File: apps/server/src/ws/state.ts
import type { ClientGameView, ClientRoundState, PlayerId } from '@game/domain';
import type { ServerRoom } from '../rooms/RoomRegistry.js';

export function buildClientGameView(room: ServerRoom, playerId?: PlayerId): ClientGameView {
  const { gameState } = room;
  const round = gameState.roundState;
  const { minPlayers, maxPlayers, roundCount } = gameState.config;

  const clientRound: ClientRoundState | null = round
    ? {
        roundIndex: round.roundIndex,
        cardsPerPlayer: round.cardsPerPlayer,
        trumpSuit: round.trumpSuit,
        trumpCard: round.trumpCard,
        trumpBroken: round.trumpBroken,
        trickInProgress: round.trickInProgress,
        completedTricks: round.completedTricks,
        bids: round.bids,
        dealerPlayerId: round.dealerPlayerId,
        startingPlayerId: round.startingPlayerId,
      }
    : null;

  const lobby = room.lobby
    ? {
        readyState: Object.fromEntries(
          Object.entries(room.lobby.readyState).map(([playerId, state]) => [playerId, { ...state }]),
        ),
        overrideReadyRequirement: room.lobby.overrideReadyRequirement,
      }
    : undefined;

  return {
    gameId: gameState.gameId,
    phase: gameState.phase,
    players: gameState.players,
    cumulativeScores: gameState.cumulativeScores,
    roundSummaries: gameState.roundSummaries,
    you: playerId,
    hand: playerId ? gameState.playerStates[playerId]?.hand ?? [] : undefined,
    round: clientRound,
    config: {
      minPlayers,
      maxPlayers,
      roundCount,
    },
    joinCode: room.joinCode,
    isPublic: room.isPublic,
    lobby,
  };
}

File: apps/mcts-ai/src/engine/state.py
from typing import List, Optional, Dict, Literal
from pydantic import BaseModel
from .cards import Card, Suit

# Player Types
PlayerId = str

class PlayerProfile(BaseModel):
    displayName: str
    avatarSeed: str
    color: str
    userId: Optional[str] = None

class PlayerInGame(BaseModel):
    playerId: PlayerId
    seatIndex: Optional[int]
    profile: Optional[PlayerProfile] = None  # Make optional for tests
    status: Literal['active', 'disconnected', 'left'] = 'active' # Default for tests
    isBot: bool
    spectator: bool

class ServerPlayerState(BaseModel):
    playerId: PlayerId
    hand: List[Card]
    tricksWon: int
    bid: Optional[int]
    roundScoreDelta: int = 0

# Game State Types
class TrickPlay(BaseModel):
    playerId: PlayerId
    card: Card
    order: int

class TrickState(BaseModel):
    trickIndex: int
    leaderPlayerId: PlayerId
    ledSuit: Optional[Suit]
    plays: List[TrickPlay]
    winningPlayerId: Optional[PlayerId]
    winningCardId: Optional[str]
    completed: bool

class RoundState(BaseModel):
    roundIndex: int
    cardsPerPlayer: int
    roundSeed: str
    trumpCard: Optional[Card]
    trumpSuit: Optional[Suit]
    trumpBroken: bool
    bids: Dict[PlayerId, Optional[int]]
    biddingComplete: bool
    trickInProgress: Optional[TrickState]
    completedTricks: List[TrickState]
    dealerPlayerId: Optional[PlayerId]
    startingPlayerId: Optional[PlayerId]
    deck: List[Card]
    remainingDeck: List[Card]

class GameConfig(BaseModel):
    gameId: str
    sessionSeed: str
    roundCount: int
    minPlayers: int
    maxPlayers: int

class GameState(BaseModel):
    gameId: str
    config: GameConfig
    phase: Literal['LOBBY', 'BIDDING', 'PLAYING', 'SCORING', 'COMPLETED']
    players: List[PlayerInGame]
    playerStates: Dict[PlayerId, ServerPlayerState]
    roundState: Optional[RoundState]
    cumulativeScores: Dict[PlayerId, int]
